src/Mapper.php

```diff
@@ -1,11 +1,156 @@
 <?php
 
-namespace Tnapf\JsonMapper\Mapper;
+namespace Tnapf\JsonMapper;
+
+use ReflectionAttribute;
+use ReflectionClass;
+use ReflectionProperty;
+use Tnapf\JsonMapper\Attributes\AnyArray;
+use Tnapf\JsonMapper\Attributes\AnyType;
+use Tnapf\JsonMapper\Attributes\CaseConversionInterface;
+use Tnapf\JsonMapper\Attributes\BaseType;
+use Tnapf\JsonMapper\Attributes\BoolType;
+use Tnapf\JsonMapper\Attributes\FloatType;
+use Tnapf\JsonMapper\Attributes\IntType;
+use Tnapf\JsonMapper\Attributes\ObjectArrayType;
+use Tnapf\JsonMapper\Attributes\ObjectType;
+use Tnapf\JsonMapper\Attributes\StringType;
 
 class Mapper implements MapperInterface
 {
+    protected ?CaseConversionInterface $caseConversion = null;
+    protected object $instance;
+    protected ReflectionClass $reflection;
+
+    /**
+     * @var BaseType[]
+     */
+    protected array $attributes;
+
+    protected string $class;
+    protected array $data;
+
     public function map(string $class, array $data): object
     {
-        $instance = new $class();
+        $instance = (new self());
+        $instance->class = $class;
+        $instance->data = $data;
+        $instance->reflection = new ReflectionClass($class);
+
+        if ($attributes = $instance->reflection->getAttributes(CaseConversionInterface::class, ReflectionAttribute::IS_INSTANCEOF)) {
+            if (count($attributes) > 1) {
+                throw new MapperException("{$class} has more than one case conversion attribute");
+            }
+
+            $instance->caseConversion = $attributes[0]->newInstance();
+        }
+
+        $instance->instance = new $class();
+
+        return $instance->doMapping();
+    }
+
+    protected function convertNameToCase(string $name): string
+    {
+        return $this?->caseConversion?->convertToCase($name) ?? $name;
+    }
+
+    protected function convertNameFromCase(string $name): string
+    {
+        return $this?->caseConversion?->convertFromCase($name) ?? $name;
+    }
+
+    protected function doMapping(): object
+    {
+        $this->fillPropertyAttributes();
+
+        foreach ($this->attributes as $types) {
+            $attribute = $types[0];
+            $data = $this->data[$this->convertNameFromCase($attribute->name)] ?? null;
+
+            if ($data === null) {
+                if ($attribute->nullable) {
+                    continue;
+                }
+
+                throw new MapperException("Property {$attribute->name} is not nullable");
+            }
+
+            $validType = false;
+
+            foreach ($types as $attribute) {
```

Comment:
This is weird. Almost if these two loops should be separated, because two things happen here that don't collide: initial data modification and data type validation. Up to you tho.

Reply:
Did it in one loop for speed. Probably could create methods inside ObjectType & ObjectArrayType for the data modification but at the same time there isn't enough there to make it plausible.